<a href="https://colab.research.google.com/github/adhang/data-science-digitalskola/blob/update/13.%20Advanced%20Statistics/Learn%20-%20Advanced%20Statistics.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Advanced Statistics
Author: Adhang Muntaha Muhammad

[![LinkedIn](https://img.shields.io/badge/linkedin-0077B5?style=for-the-badge&logo=linkedin&logoColor=white&link=https://www.linkedin.com/in/adhangmuntaha/)](https://www.linkedin.com/in/adhangmuntaha/)
[![GitHub](https://img.shields.io/badge/github-121011?style=for-the-badge&logo=github&logoColor=white&link=https://github.com/adhang)](https://github.com/adhang)
[![Kaggle](https://img.shields.io/badge/kaggle-20BEFF?style=for-the-badge&logo=kaggle&logoColor=white&link=https://www.kaggle.com/adhang)](https://www.kaggle.com/adhang)
[![Tableau](https://img.shields.io/badge/tableau-E97627?style=for-the-badge&logo=tableau&logoColor=white&link=https://public.tableau.com/app/profile/adhang)](https://public.tableau.com/app/profile/adhang)
___
**Contents**
1. Sampling
  - Simple Random Sampling
  - Systematic Sampling
  - Stratified Sampling
2. Hypothesis Testing
  - Z-test
  - T-test
  - Chi-square
  - Two Test Sample
<br>

**Notes for Z-test**
- For some reason, when I run the z-test code, I got an error (`TypeError: 'numpy.float64' object is not callable`)
- To solve this problem, I restarted my kernel and re-run my notebook
  - Runtime -> Restart runtime -> Run all
  <br>or
  - Runtime -> Restart and run all
- I still can't figure out the root cause of this error
<br>

**Notes for T-test**
- When I run the t-test code, I got an error (`TypeError: ttest_1samp() got an unexpected keyword argument 'alternative'`)
- I need to upgrade the Scipy library to solve this problem because this parameter is only available on Scipy version 1.6.0 or higher

# Importing Libraries

In [1]:
!pip install statsmodels

!pip install --upgrade scipy

In [2]:
import pandas as pd
import numpy as np

from statsmodels.stats.weightstats import ztest
import scipy.stats as stats

/usr/local/lib/python3.7/dist-packages/statsmodels/tools/_testing.py:19: FutureWarning: pandas.util.testing is deprecated. Use the functions in the public API at pandas.testing instead.
  import pandas.util.testing as tm


# Reading Dataset
Transformer = trafo listrik

In [3]:
path = 'https://raw.githubusercontent.com/densaiko/data_science_learning/main/dataset/Transformer.csv'

data = pd.read_csv(path)
data.head()

Unnamed: 0  Hydrogen  Oxigen  ...  Health index  Life expectation        event
0           0      2845    5860  ...          95.2              19.0  Not Working
1           1     12886      61  ...          85.5              19.0  Not Working
2           2      2820   16400  ...          85.3              19.0  Not Working
3           3      1099      70  ...          85.3               6.0  Not Working
4           4      3210    3570  ...          85.2               6.0  Not Working

[5 rows x 18 columns]

# Sampling

## Simple Random Sampling

In [ ]:
sample_random = data.sample(n=20)
sample_random.shape

(20, 18)

In [ ]:
sample_random.head()

Unnamed: 0  Hydrogen  Oxigen  ...  Health index  Life expectation        event
110         110        16    7340  ...          48.2               6.5  Not Working
122         122        24    1330  ...          38.3              19.0  Not Working
145         145         4   17000  ...          38.3              19.0  Not Working
289         289        22     399  ...          13.4              51.0      Working
457         457        16    2020  ...          13.4              51.0      Working

[5 rows x 18 columns]

## Systematic Sampling

In [ ]:
index = np.arange(0, len(data), 4)
sample_systematic = data.iloc[index]
sample_systematic.head()

Unnamed: 0  Hydrogen  Oxigen  ...  Health index  Life expectation        event
0            0      2845    5860  ...          95.2              19.0  Not Working
4            4      3210    3570  ...          85.2               6.0  Not Working
8            8        16    2470  ...          72.8               6.0  Not Working
12          12       254     592  ...          61.3               6.0  Not Working
16          16      6620    5250  ...          60.5              19.0  Not Working

[5 rows x 18 columns]

## Stratified Sampling
Just like filtering data

In [ ]:
data[data['event'] == 'Not Working']

Unnamed: 0  Hydrogen  Oxigen  ...  Health index  Life expectation        event
0             0      2845    5860  ...          95.2              19.0  Not Working
1             1     12886      61  ...          85.5              19.0  Not Working
2             2      2820   16400  ...          85.3              19.0  Not Working
3             3      1099      70  ...          85.3               6.0  Not Working
4             4      3210    3570  ...          85.2               6.0  Not Working
..          ...       ...     ...  ...           ...               ...          ...
230         230        50     205  ...          13.4              51.0  Not Working
231         231        49     352  ...          13.4              51.0  Not Working
232         232        49     398  ...          13.4              51.0  Not Working
233         233        48     245  ...          13.4              51.0  Not Working
234         234        48    1330  ...          13.4              51.0  Not Working

[235 rows x 18 columns]

In [ ]:
data[data['Health index'] <= 80]

Unnamed: 0  Hydrogen  Oxigen  ...  Health index  Life expectation        event
5             5     13500     343  ...          75.6               6.0  Not Working
6             6     10200   11900  ...          75.6               6.0  Not Working
7             7         3   15459  ...          73.2              19.0  Not Working
8             8        16    2470  ...          72.8               6.0  Not Working
9             9       488   11861  ...          68.0               6.0  Not Working
..          ...       ...     ...  ...           ...               ...          ...
465         465        15     227  ...          13.4              51.0      Working
466         466        15     334  ...          13.4              51.0      Working
467         467        15    1280  ...          13.4               6.5      Working
468         468        15     169  ...          13.4              51.0      Working
469         469        15     308  ...          13.4              51.0      Working

[465 rows x 18 columns]

# Hypothesis Testing

## Z-test
For this case, I will use `CO2` attribute (one sample case).

In [ ]:
overall_CO2_avg = data['CO2'].mean()
overall_CO2_std = data['CO2'].std()

print(f'The overall CO2 average is\t\t: {round(overall_CO2_avg,2)}')
print(f'The overall CO2 standard deviation is\t: {round(overall_CO2_std,2)}')

The overall CO2 average is		: 1816.41
The overall CO2 standard deviation is	: 2256.79


Our hypotheses:
- H0: The average CO2 produced when the transformer works is greater than the average CO2 produced (1816.41)
- H1: The average CO2 produced when the transformer works is less than the average CO2 produced (1816.41)
<br>

And suppose:
- We will use 35 sample using random sampling
- Confidence iterval is 95%, therefore alpha is 0.05 (5%)

In [ ]:
# stratified sampling
data_working = data[data['event'] == 'Working']

# random sampling 
sample_working_CO2 = data_working['CO2'].sample(n=35)

# z-test
ztest, pvalue = ztest(x1=sample_working_CO2, x2=None, value=overall_CO2_avg, alternative='larger')
print(f'ztest: {round(ztest,2)}, p-value: {round(pvalue,2)}')

# confidence interval and alpha
confidence_interval = 0.95
alpha = 1 - confidence_interval

# evaluating the hypothesis
print('\n')
if pvalue < alpha:
  print('Reject the hypothesis')
  print(f'Since p-value ({round(pvalue,2)}) < alpha ({round(alpha,2)})')
else:
  print('Accept the hypothesis')
  print(f'Since p-value ({round(pvalue,2)}) > alpha ({round(alpha,2)})')

ztest: 0.39, p-value: 0.35


Accept the hypothesis
Since p-value (0.35) > alpha (0.05)


## T-test
For this case, I will use `Oxigen` attribute (one sample case). Yes, the column name is `Oxigen`, not `Oxygen` or `O2`. I just follow the dataset.

In [4]:
overall_O2_avg = data['Oxigen'].mean()
overall_O2_std = data['Oxigen'].std()

print(f'The overall O2 average is\t\t: {round(overall_O2_avg,2)}')
print(f'The overall O2 standard deviation is\t: {round(overall_O2_std,2)}')

The overall O2 average is		: 8357.37
The overall O2 standard deviation is	: 14164.23


Our hypotheses:
- H0: The average O2 produced when the transformer works is less than the average O2 produced (8357.37)
- H1: The average O2 produced when the transformer works is greater than the average O2 produced (8357.37)
<br>

And suppose:
- We will use 25 sample using random sampling
- Confidence iterval is 95%, therefore alpha is 0.05 (5%)

In [5]:
# stratified sampling
data_working = data[data['event'] == 'Working']

# random sampling 
sample_working_O2 = data_working['Oxigen'].sample(n=25)

# t-test
ttest, pvalue = stats.ttest_1samp(sample_working_O2, overall_O2_avg, alternative='less')
print(f'ttest: {round(ttest,2)}, p-value: {round(pvalue,2)}')

# confidence interval and alpha
confidence_interval = 0.95
alpha = 1 - confidence_interval

# evaluating the hypothesis
print('\n')
if pvalue < alpha:
  print('Reject the hypothesis')
  print(f'Since p-value ({round(pvalue,2)}) < alpha ({round(alpha,2)})')
else:
  print('Accept the hypothesis')
  print(f'Since p-value ({round(pvalue,2)}) > alpha ({round(alpha,2)})')

ttest: -0.27, p-value: 0.39


Accept the hypothesis
Since p-value (0.39) > alpha (0.05)


## Chi-square
Suppose we want to know if there is a relationship between `CO`, `CO2`, `Hydrogen`, and `Methane` attributes based on the `event` attribute.
<br><br>
This is called chi-square test for independence, also called Pearson's chi-square test or the chi-square test of association. In this test we are testing the association between two or more variable ([source](https://stats.stackexchange.com/questions/110718/chi-squared-test-with-scipy-whats-the-difference-between-chi2-contingency-and)).
<br><br>
And suppose:
- Confidence iterval is 95%, therefore alpha is 0.05 (5%)

In [ ]:
chi = data[['event', 'CO', 'CO2', 'Hydrogen', 'Methane']]
chi = chi.set_index('event')
chi.head()

CO   CO2  Hydrogen  Methane
event                                    
Not Working   32  1344      2845     7406
Not Working   83   864     12886      877
Not Working  257  1080      2820      144
Not Working  184  1402      1099      545
Not Working  360  2130      3210      160

In [ ]:
# chi-square
chi2, pvalue, dof, expected = stats.chi2_contingency(chi)
print(f'chi2: {round(chi2,2)}, p-value: {round(pvalue,2)}')

# confidence interval and alpha
confidence_interval = 0.95
alpha = 1 - confidence_interval

# evaluating the hypothesis
print('\n')
if pvalue < alpha:
  print('Reject the hypothesis')
  print(f'Since p-value ({round(pvalue,2)}) < alpha ({round(alpha,2)})')
else:
  print('Accept the hypothesis')
  print(f'Since p-value ({round(pvalue,2)}) > alpha ({round(alpha,2)})')

chi2: 1094042.93, p-value: 0.0


Reject the hypothesis
Since p-value (0.0) < alpha (0.05)


## Two Test Sample
For this case, I will use `Methane` attribute (two sample case).
- Case 1: The methane produced when the transformer is working
- Case 2: The methane produced when the transformer is not working

In [ ]:
overall_methane_avg = data['Methane'].mean()
overall_methane_std = data['Methane'].std()

print(f'The overall Methane average is\t\t\t: {round(overall_methane_avg,2)}')
print(f'The overall Methane standard deviation is\t: {round(overall_methane_std,2)}')

The overall Methane average is			: 79.7
The overall Methane standard deviation is	: 489.32


Our hypotheses:
- H0: The average methane produced when the transformer is working is the same as when the transformer is not working
- H1: The average methane produced when the transformer is working is not the same as when the transformer is not working

### Using Z-test
Suppose:
- We will use 50 sample using random sampling
- Confidence iterval is 95%, therefore alpha is 0.05 (5%)

In [ ]:
# stratified sampling
data_working = data[data['event'] == 'Working']
data_not_working = data[data['event'] == 'Not Working']

# random sampling 
sample_working_methane = data_working['Methane'].sample(n=50)
sample_not_working_methane = data_not_working['Methane'].sample(n=50)

# z-test
ztest, pvalue = ztest(x1=sample_working_methane, x2=sample_not_working_methane, value=0, alternative='two-sided')
print(f'ztest: {round(ztest,2)}, p-value: {round(pvalue,2)}')

# confidence interval and alpha
confidence_interval = 0.95
alpha = 1 - confidence_interval

# evaluating the hypothesis
print('\n')
if pvalue < alpha:
  print('Reject the hypothesis')
  print(f'Since p-value ({round(pvalue,2)}) < alpha ({round(alpha,2)})')
else:
  print('Accept the hypothesis')
  print(f'Since p-value ({round(pvalue,2)}) > alpha ({round(alpha,2)})')

ztest: -2.2, p-value: 0.03


Reject the hypothesis
Since p-value (0.03) < alpha (0.05)


### Using T-test
Suppose:
- We will use 25 sample using random sampling
- Confidence iterval is 95%, therefore alpha is 0.05 (5%)

In [6]:
# stratified sampling
data_working = data[data['event'] == 'Working']
data_not_working = data[data['event'] == 'Not Working']

# random sampling 
sample_working_methane = data_working['Methane'].sample(n=25)
sample_not_working_methane = data_not_working['Methane'].sample(n=25)

# t-test
ttest, pvalue = stats.ttest_ind(a=sample_working_methane, b=sample_not_working_methane, equal_var=True)
print(f'ttest: {round(ttest,2)}, p-value: {round(pvalue,2)}')

# confidence interval and alpha
confidence_interval = 0.95
alpha = 1 - confidence_interval

# evaluating the hypothesis
print('\n')
if pvalue < alpha:
  print('Reject the hypothesis')
  print(f'Since p-value ({round(pvalue,2)}) < alpha ({round(alpha,2)})')
else:
  print('Accept the hypothesis')
  print(f'Since p-value ({round(pvalue,2)}) > alpha ({round(alpha,2)})')

ttest: -1.69, p-value: 0.1


Accept the hypothesis
Since p-value (0.1) > alpha (0.05)
